pkg/mbw/monitor/umc.go

```diff
@@ -0,0 +1,148 @@
+			// init read umc
+			ctl = UMC_PERF_BASE + uint32(umc)*UMC_CH_SIZE + m.PMU.UMC.CtlBase + 0*m.PMU.UMC.CtlSize
+			pci.WriteSMNApp(m.PMU.SktIOHC[skt], ctl, UMC_CTL_CAS_RD)
+
+			// init write umc
+			ctl = UMC_PERF_BASE + uint32(umc)*UMC_CH_SIZE + m.PMU.UMC.CtlBase + 1*m.PMU.UMC.CtlSize
+			pci.WriteSMNApp(m.PMU.SktIOHC[skt], ctl, UMC_CTL_CAS_WR)
+
+			ctl = UMC_PERF_BASE + uint32(umc)*UMC_CH_SIZE + UMC_CLKCTL_BASE
+			pci.WriteSMNApp(m.PMU.SktIOHC[skt], ctl, 0)
+
+			general.Infof("stop UMC monitoring on socket %d and umc %d", skt, umc)
+		}
+	}
+
+	general.Infof("stop monitoring the UMC")
+}
+
+func (m MBMonitor) ClearUMCMonitor() {
+	var ctl uint32
+	for skt := 0; skt < m.KatalystMachineInfo.MachineInfo.NumSockets; skt++ {
+		for umc := 0; umc < m.PMU.UMC.NumPerSocket; umc++ {
+			ctl = UMC_PERF_BASE + uint32(umc)*UMC_CH_SIZE + UMC_CLKCTL_BASE
+			pci.WriteSMNApp(m.PMU.SktIOHC[skt], ctl, UMC_CLKCTL_G_EN)
+
+			general.Infof("clear UMC setting on socket %d and umc %d", skt, umc)
+		}
+	}
+
+	general.Infof("all UMC settings has been cleared")
+}
+
+// read the UMC by pci iohc dev
+func (m MBMonitor) ReadUMCCounter(iohc *pci.PCIDev, umc_idx, ctl_idx uint32) uint64 {
+	var val1, val2 uint64
+	var offset uint32
+
+	offset = UMC_PERF_BASE + umc_idx*UMC_CH_SIZE + m.PMU.UMC.CtrLowBase + ctl_idx*m.PMU.UMC.CtlSize
+	val1 = uint64(pci.ReadSMNApp(iohc, offset))
+
+	offset = UMC_PERF_BASE + umc_idx*UMC_CH_SIZE + m.PMU.UMC.CtrHighBase + ctl_idx*m.PMU.UMC.CtlSize
+	val2 = uint64(pci.ReadSMNApp(iohc, offset))
+
+	return ((val2 << 32) + val1) & UMC_PMC_MASK
+}
+
+// collect the memory-bandwidth for each package by reading UMC
+func (m MBMonitor) ReadPackageUMC() {
+	var cas_rd, cas_wr uint64 = 0, 0
+
+	for skt := 0; skt < m.KatalystMachineInfo.MachineInfo.NumSockets; skt++ {
+		for umc := 0; umc < m.PMU.UMC.NumPerSocket; umc++ {
+			cas_rd += m.ReadUMCCounter(m.PMU.SktIOHC[skt], uint32(umc), UMC_IOHC_READ)
+			cas_wr += m.ReadUMCCounter(m.PMU.SktIOHC[skt], uint32(umc), UMC_IOHC_WRITE)
+
+			// general.Infof("get the umc reading on socket %d and umc %d: (%d, %d)", skt, umc, cas_rd, cas_wr)
+			// accumulate the UMC result for each package (i.e. physical Numa)
+			if (umc+1)%m.PMU.UMC.NumPerPackage == 0 {
+				pkgIdx := (skt*m.PMU.UMC.NumPerSocket + umc) / m.PMU.UMC.NumPerPackage
+				m.MemoryBandwidth.Packages[pkgIdx].RMB_Delta = utils.Delta(48, cas_rd, m.MemoryBandwidth.Packages[pkgIdx].RMB)
+				m.MemoryBandwidth.Packages[pkgIdx].WMB_Delta = utils.Delta(48, cas_wr, m.MemoryBandwidth.Packages[pkgIdx].WMB)
+				m.MemoryBandwidth.Packages[pkgIdx].Total = utils.PMUToMB(m.MemoryBandwidth.Packages[pkgIdx].WMB_Delta+m.MemoryBandwidth.Packages[pkgIdx].RMB_Delta, m.Interval)
+
+				m.MemoryBandwidth.Packages[pkgIdx].RMB = cas_rd
+				m.MemoryBandwidth.Packages[pkgIdx].WMB = cas_wr
+				// general.Infof("get the mb result for physical node %d: (%d, %d)", (skt*m.UMC.NumPerSocket+umc)/m.UMC.NumPerPackage, cas_rd, cas_wr)
```

Comment:
either remove or uncomment it. same to line 131

Reply:
ok will remove all the commented log info lines

Reply:
ok will remove commented out log lines